### PR TITLE
An example on learnable image resizing

### PR DESCRIPTION
**Pull request by sayakpaul**
Adaptive image resizing is being used for some time. It has either been used in the form of progressive resizing or in more explicit adaptive forms such as in EfficientNetV2. 

This example, however, introduces an example that focuses on this question as investigated [here](https://arxiv.org/pdf/2103.09950v1.pdf) - "how to optimally learn representations for a given image resolution?" The improvements are quite nice and I hope it will be useful for the community to improve their vision models.

### CODE REVIEW

File: examples/vision/learnable_resizer.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+plt.figure(figsize=(10, 10))
+for i, image in enumerate(sample_images[:9]):
+    ax = plt.subplot(3, 3, i + 1)
+    image = tf.image.convert_image_dtype(image, tf.float32)
+    resized_image = learned_resizer(image[None, ...])
+    plt.imshow(resized_image.numpy().squeeze())
+    plt.axis("off")
+
+"""
+As we can see the visuals of the images have improved with training. Additionally, you
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:242]: Tip: for accessibility, I'd recommend avoiding using "as we can see":

```suggestion
The plots show that the images have improved with training. Additionally, you
```

File: examples/vision/learnable_resizer.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+can find this repository that shows the benefits of using the resizing module. Below is a
+comparison:
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:244]:
```suggestion
Suggestion:

- If you're referring to the your `Learnable-Image-Resizing` GitHub repo, let's refer to it here (as well as later, as you have already). Otherwise, it's unclear what "this repository" means here.

Maybe we could rephrase this as follows:

```suggestion
can use the `sayakpaul/Learnable-Image-Resizing` repository that shows the
benefits of using the resizing module—here is a comparison table:
```

File: examples/vision/learnable_resizer.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+|           Model           	| Number of  parameters (Million) 	| Top-1 accuracy 	|
+|:-------------------------:	|:-------------------------------:	|:--------------:	|
+|   With learnable resizer  	|             7.051717            	|      52.02     	|
+| Without learnable resizer 	|             7.039554            	|      50.3      	|
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:249]: Suggestion:

- "the learnable resizer" (with the "the" (definite article))

File: examples/vision/learnable_resizer.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+Note the above-reported models were trained for 10 epochs on 90% of the training set of
+Cats-vs-Dogs unlike this example. Also, note that the increase in the number of
+parameters due to the resizing module is very negligible. You can reproduce these results
+from [this repository](https://github.com/sayakpaul/Learnable-Image-Resizing). To ensure
+that the improvement in the performance is not due to stochasticity, the models were
+trained using the same initial random weights.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:256]: Suggestion: refactoring (with 80 char per line limit)

```suggestion
Note: Unlike this example, the comparable models in the above table were trained
for 10 epochs on 90% of the training set of the Cats-vs-Dogs dataset. Notice the
increase in the number of parameters—this is because the resizing module is
negligible. You can reproduce these results by using
[`sayakpaul/Learnable-Image-Resizing`](https://github.com/sayakpaul/Learnable-Image-Resizing)
repository. And, to make sure that the improvement in the performance is not due
to stochasticity, the models were trained using the same initial random weights.
```

File: examples/vision/learnable_resizer.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+combination of natural and stylized images. It might be interesting to investigate if
+this learnable resizing module could achieve something similar as the outputs seem to
+discard the texture information. 
+* Through a set of experiments, the authors also verify if the performance improvement is
+***not*** solely due to the increase in the number of model parameters. You are
+encouraged to check those out in the
+[original paper](https://arxiv.org/abs/2103.09950v1). 
+* There is another closely related topic on adaptive image resizing that attempts to
+resize images/feature maps adaptively during training. [EfficientV2](https://arxiv.org/pdf/2104.00298)
+uses this idea.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:272]:
```suggestion
- Note: "et al." <- "et. al"
- Included the link to the first paper, so the user doesn't have to search for it again
- Included the authors of the second and third papers
- "You may find it interesting" <- "It might be interesting" (more sort of active voice this way, talking to the user)
- 80 char per line

```
* To impose shape bias inside the vision models, 
[Geirhos et al.](https://arxiv.org/abs/1811.12231) trained them with a 
combination of natural and stylized images. You may find it interesting to 
investigate if this learnable resizing module could achieve something similar, 
as the outputs appear to discard the texture information. 
* Through a set of experiments, the authors also verify if the performance
improvement is _not_ solely due to the increase in the number of model
parameters. (Check out the [original
paper](https://arxiv.org/abs/2103.09950v1) by Telebi et al.).
* There is another closely related topic on adaptive image resizing that
attempts to resize images/feature maps adaptively during training—
[EfficientV2](https://arxiv.org/pdf/2104.00298) by Tan & Le uses this idea. 
```

File: examples/vision/learnable_resizer.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+    validation_ds.map(preprocess_dataset, num_parallel_calls=AUTO)
+    .batch(BATCH_SIZE)
+    .prefetch(AUTO)
+)
+
+"""
+## Define the learnable resizer utilities
+
+The figure below (courtesy of the original paper) presents the structure of the learnable
+resizing module:
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:115]: Suggestion:

- There are at least 2 papers here, so to make it clear which "original paper" this was taken from, let's be more specific: it's from Learning to Resize Images for Computer Vision Tasks by Talebi & Milanfar (2021)


```suggestion
Below is the structure of the learnable resizing module (source: Learning to
Resize Images for Computer Vision Tasks by Talebi & Milanfar (2021)):
```

File: examples/vision/learnable_resizer.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+"""
+For this example, we will use the bilinear interpolation but the learnable image resizer
+module is not dependent on any specific interpolation method. We can use other ones (such
+as bicubic) as well. 
+"""
+
+"""
+## Load and prepare the dataset
+
+For this example, we will only use 40% of the total training dataset.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:82]: Suggestion:

```suggestion
In this example, we will only use 40% of the training set from Kaggle Cats and Dogs.
```

(Full name: Kaggle Cats and Dogs - https://www.microsoft.com/en-us/download/details.aspx?id=54765)

File: examples/vision/learnable_resizer.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+AUTO = tf.data.AUTOTUNE
+BATCH_SIZE = 64
+EPOCHS = 5
+ALPHA = 0.2
+
+"""
+For this example, we will use the bilinear interpolation but the learnable image resizer
+module is not dependent on any specific interpolation method. We can use other ones (such
+as bicubic) as well.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:76]: Suggestion: refactor, 80 char limit 

```suggestion
In this example, we will use the bilinear interpolation but the learnable image
resizer module is not dependent on any specific interpolation method. We can
also use others, such as bicubic.
```

File: examples/vision/learnable_resizer.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+
+"""
+## Define hyperparameters
+"""
+
+"""
+In order to facilitate mini-batch learning, we need to have a fixed shape for the images
+inside a given batch. This is why an initial resizing is required. We first resize all
+the images to (300x300) shape and then learn their optimal representation for the
+(224x224) resolution.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:61]: Suggestion: refactor, 80 char limit

```suggestion
In order to facilitate mini-batch learning, we need to have a fixed shape for
the images inside a given batch. To do that, we need to carry out initial
resizing: we first resize all the images to (300x300) shape and then learn their
optimal representation for the (224x224) resolution.
```

File: examples/vision/learnable_resizer.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+the following question: 
+
+**For a given image resolution and a model, how to best resize the given images?**
+
+As shown in the paper, this idea helps to consistently improve the performance of the
+common vision models (pre-trained on ImageNet-1k) like DenseNet121, ResNet50,
+MobileNetV2, and EfficientNets. In this example, we will implement the learnable image
+resizing module as proposed in the paper and demonstrate that on the
+[Cast-Vs-Dogs dataset](https://www.microsoft.com/en-us/download/details.aspx?id=54765)
+using the [DenseNet121](https://arxiv.org/abs/1608.06993) model.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:33]: Very cool intro.

Suggestion:

- Break into more paragraphs for improved readability
- Add the first paper's name (ImageNet-trained CNNs are biased towards texture; increasing shape bias improves accuracy and robustness)
- "can be biased" <- "are biased" (nothing is 100% certain?)
- "DenseNet-121", "ResNet-50" with a dash
- "Kaggle Cats and Dogs" <- "Cast-Vs-Dogs" (typo)
- "DenseNet-121 architecture" <- "...model" (https://www.tensorflow.org/api_docs/python/tf/keras/applications/DenseNet121)
- Resolution (Number x Number)
- 80 char per line

```suggestion
It is a common belief that if we constrain vision models to perceive things as
humans do, their performance can be improved. For example, in
[ImageNet-trained CNNs are biased towards texture; increasing shape bias improves accuracy and robustness](https://arxiv.org/abs/1811.12231)
, Geirhos et al. showed that vision models pre-trained on the ImageNet 1000
dataset can be biased toward texture, whereas human beings mostly use the shape 
descriptor to develop a common perception. 

_However, does this belief always apply especially when it comes to improving
the performance of vision models?_ It turns out it may not always be the case.
When training vision models, it is common to resize images to a lower dimension
(such as (224 x 224), (299 x 299), etc.) to allow mini-batch learning and keep 
up with the compute limitations.

For this step, we generally use image resizing methods, such as **bilinear
interpolation**, while the resized images do not lose much of their perceptual
character to the human eyes. 

In [Learning to Resize Images for Computer Vision Tasks](https://arxiv.org/abs/2103.09950v1)
, Talebi et. al demonstrate that if we try to optimize the perceptual quality
of the images for the vision models rather than the human eyes, their
performance can further be improved. They investigate the following question: _for a given image resolution and a model, how can we best resize the given
images?_ As shown in the paper, this idea helps to consistently improve the
performance of the common vision models (pre-trained on ImageNet-1k), such as
DenseNet-121, ResNet-50, MobileNetV2, and EfficientNets. 

In this example, we will implement the learnable image resizing module as
proposed in the paper and demonstrate that on the
[Kaggle Cats and Dogs dataset](https://www.microsoft.com/en-us/download/details.aspx?id=54765)
using the [DenseNet-121](https://arxiv.org/abs/1608.06993) architecture.
```

File: examples/vision/learnable_resizer.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+    x = conv_block(x, 16, 3, 1)
+    x = conv_block(x, 16, 3, 1, activation=None)
+    return layers.Add()([inputs, x])
+
+
+def learnable_resizer(
+    inputs, filters=16, num_res_blocks=1, interpolation=INTERPOLATION
+):
+
+    # We first do a naive resizing.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:140]:
```suggestion
    # First, perform naive resizing.
```

File: examples/vision/learnable_resizer.py [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+    naive_resize = layers.experimental.preprocessing.Resizing(
+        *TARGET_DIM, interpolation=interpolation
+    )(inputs)
+
+    # First conv block without Batch Norm.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:145]:
```suggestion
    # First convolution block without batch normalization.
```

File: examples/vision/learnable_resizer.py [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+    x = layers.Conv2D(filters=filters, kernel_size=7, strides=1, padding="same")(inputs)
+    x = layers.LeakyReLU(ALPHA)(x)
+
+    # Second conv block with Batch Norm.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:149]:
```suggestion
    # Second convolution block with batch normalization.
```

File: examples/vision/learnable_resizer.py [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+    x = layers.Conv2D(filters=filters, kernel_size=1, strides=1, padding="same")(x)
+    x = layers.LeakyReLU(ALPHA)(x)
+    x = layers.BatchNormalization()(x)
+
+    # Intermediate resizing as bottleneck.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:154]:
```suggestion
    # Intermediate resizing as a bottleneck.
```

File: examples/vision/learnable_resizer.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+    final_resize = layers.Add()([naive_resize, x])
+
+    return final_resize
+
+
+"""
+## Visualize the outputs of the learnable resizing module
+
+Here, we visualize how the resized images would look like after being passed through the
+random weights of the resizer.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:183]:
```suggestion
Let's visualize how the resized images would look like after being passed
through the random weights of the resizer. 
```

**Review by 8bitmp3 — COMMENTED** [review]
(Another) excellent example @sayakpaul 

I left a few suggestions. PTAL. Thank you 👍 👍 👍

File: examples/vision/learnable_resizer.py [hunk 16]
```diff
@@ -263,11 +263,13 @@ def get_model():
 combination of natural and stylized images. It might be interesting to investigate if
 this learnable resizing module could achieve something similar as the outputs seem to
 discard the texture information. 
+* The resizer module can handle arbitrary resolutions and aspect ratios which is very
+important for tasks like object detection and segmentation. 
 * Through a set of experiments, the authors also verify if the performance improvement is
 ***not*** solely due to the increase in the number of model parameters. You are
 encouraged to check those out in the
 [original paper](https://arxiv.org/abs/2103.09950v1).
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:271]: I highly recommend spelling out the name of the paper here. There are at least 2 papers in this example. Therefore, "original paper" is kind of confusing and the user would have to click on it to find out what the reference is about.

Comment (sayakpaul) [reply examples/vision/learnable_resizer.py:274]: The name of the paper is too long for that purpose IMO. Hence, I think mentioning it one/two times is fine.

File: examples/vision/learnable_resizer.py [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+
+"""
+## Define hyperparameters
+"""
+
+"""
+In order to facilitate mini-batch learning, we need to have a fixed shape for the images
+inside a given batch. This is why an initial resizing is required. We first resize all
+the images to (300x300) shape and then learn their optimal representation for the
+(224x224) resolution.
```
Comment (8bitmp3) [inline examples/vision/learnable_resizer.py:61]: Nit:

```suggestion
the images to (300 x 300) shape and then learn their optimal representation for the
(224 x 224) resolution. 
```

File: examples/vision/learnable_resizer.py (hunk 16 above)
Comment (8bitmp3) [reply examples/vision/learnable_resizer.py:274]: Yup. IMO, you should still make a distinction w.r.t what the "original paper" means for the readers.

You have:
- [ImageNet-trained CNNs are biased towards texture; increasing shape bias improves accuracy and robustness](https://arxiv.org/abs/1811.12231)
- [Learning to Resize Images for Computer Vision Tasks](https://arxiv.org/abs/2103.09950v1)

Using their long names is not an issue, in my view, as long as you help the users.

Comment (sayakpaul) [reply examples/vision/learnable_resizer.py:274]: Since at the beginning the paper's name is spelled out, it's okay to refer to it as "original paper" IMO.

Comment (8bitmp3) [reply examples/vision/learnable_resizer.py:274]: Gotcha 👍  (the Notes section (and the majority of the example) is/are based on the idea of learned resizers from _Learning to Resize Images for Computer Vision Tasks_, which is the "original paper".)

**Comment by 8bitmp3** [comment]
A couple more (hopefully last) ideas for improvement of your awesome example to help new users:

- Since the example is technically about the cool idea of learned resizers (instead of traditional ones), maybe we could mention that in the description explicitly for the less informed users and to help with search. (Keeping in mind that the title already mentions "learning to resize").

  For example:

```diff
- Description: How to optimally learn representations of images for a given resolution.
+ Description: Customized pre-processing with a learned resizer for optimal image representation learning for a given resolution.
```

- Maybe we should use "_learned_ resizer" instead of "_learnable_ resizer":
  - The literature (https://arxiv.org/pdf/1811.12231.pdf and https://arxiv.org/pdf/2103.09950v1.pdf) used in the example specifically refers to it as "learned features" or "learned resizer".
  - (Having said that, some papers, such as https://www.ecva.net/papers/eccv_2020/papers_ECCV/papers/123570647.pdf, use "learnable" 🤷‍♂️ .)

Anyway, these are just some ideas. Hope this helps.

**Comment by sayakpaul** [comment]
IMO this would overcomplicate things. So, I would just keep the things as they are for now. I think I have made efforts to describe the purpose of why would someone be interested in the learning task in the first place. On the other hand, the description you provided would break the check for its length. So, keeping that in mind (and to make a description that makes sense) I would prefer the current one.

File: examples/vision/learnable_resizer.py (hunk 10 above)
Comment (8bitmp3) [reply examples/vision/learnable_resizer.py]: > 80 char per line

(Nit: as per @ebursztein's suggestion in another example by a different user, maybe we should have lines no longer than 80 characters (instead of 90 or 100). Maybe, this can be done your future example, which are awesome as always.)

Comment (sayakpaul) [reply examples/vision/learnable_resizer.py]: Thank you!

I try to keep the characters in a line such that they are visually not problematic.

File: examples/vision/learnable_resizer.py [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+    plt.axis("off")
+
+"""
+The plot shows that the visuals of the images have improved with training. The following
+table shows the benefits of using the resizing module in comparison to using the bilinear 
+interpolation:
+
+|           Model           	| Number of  parameters (Million) 	| Top-1 accuracy 	|
+|:-------------------------:	|:-------------------------------:	|:--------------:	|
+|   With learnable resizer  	|             7.051717            	|      52.02     	|
```
Comment (fchollet) [inline examples/vision/learnable_resizer.py:251]: The resizer is a residual convnet. Isn't the improved accuracy simply a consequence of adding more layers (representational power) to the model, compared to the baseline? Maybe this is explained in the paper (didn't have time to read it). How do we argue the scientific interest of these results?

File: examples/vision/learnable_resizer.py [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+biased toward texture whereas human beings mostly use the shape descriptor to develop a
+common perception. But does this belief always apply especially when it comes to improving
+the performance of vision models? 
+
+It turns out it may not always be the case. When training vision models, it is common to
+resize images to a lower dimension ((224 x 224), (299 x 299), etc.) to allow mini-batch
+learning and also to keep up the compute limitations.  We generally make use of image
+resizing methods like **bilinear interpolation** for this step and the resized images do
+not lose much of their perceptual character to the human eyes. In [Learning to Resize
+Images for Computer Vision Tasks](https://arxiv.org/abs/2103.09950v1), Talebi et al. show
```
Comment (fchollet) [inline examples/vision/learnable_resizer.py:21]: The cited paper is a preprint from 2 month ago with 0 citations -- do we have external guarantees of its quality and importance?

File: examples/vision/learnable_resizer.py [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+"""
+
+"""
+## Notes
+
+* To impose shape bias inside the vision models, Geirhos et al. trained them with a
+combination of natural and stylized images. It might be interesting to investigate if
+this learnable resizing module could achieve something similar as the outputs seem to
+discard the texture information. 
+* The resizer module can handle arbitrary resolutions and aspect ratios which is very
```
Comment (fchollet) [inline examples/vision/learnable_resizer.py:269]: Insert line break before list

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! I have a few questions about the underlying paper, curious to hear your explanations.

File: examples/vision/learnable_resizer.py [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+
+"""
+In order to facilitate mini-batch learning, we need to have a fixed shape for the images
+inside a given batch. This is why an initial resizing is required. We first resize all
+the images to (300 x 300) shape and then learn their optimal representation for the
+(224 x 224) resolution. 
+"""
+
+INP_DIM = (300, 300)
+TARGET_DIM = (224, 224)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py:65]: Good idea. Will see to it.

File: examples/vision/learnable_resizer.py [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+):
+
+    # First, perform naive resizing.
+    naive_resize = layers.experimental.preprocessing.Resizing(
+        *TARGET_DIM, interpolation=interpolation
+    )(inputs)
+
+    # First convolution block without batch normalization.
+    x = layers.Conv2D(filters=filters, kernel_size=7, strides=1, padding="same")(inputs)
+    x = layers.LeakyReLU(ALPHA)(x)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py:147]: I can hard-code it. Since it appears multiple times, I defined it as a constant at the top.

File: examples/vision/learnable_resizer.py (hunk 18 above)
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py:251]: This is a great question. So, to show it empirically here's what the authors have done:

* Take a pre-trained model trained some size, say (224 x 224). 
* Now, first, use it to infer predictions on images resized to a lower resolution. Record the performance. 
* For the second experiment, plug in the resizer module at the top of the pre-trained model and warm-start the training. Record performance. 

Now, they argue that using the second option is better because it helps the model learn how to adjust the representations better with respect to the given resolution. I agree this might be because of an increase in the number of parameters (which I have mentioned too) more experiments would have been better to establish this claim. Experiments such as analyzing the channel-similarity index and cross-channel interaction, visualizing if the focus of the network gets better with the resizer (Grad-CAM could be used to do this), and so on.

File: examples/vision/learnable_resizer.py (hunk 19 above)
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py:21]: It's hard for me to answer. The main component that intrigued me while I was implementing this one was the idea **_that maybe learning to generate better views that are better for a model to improve its performance_**. 

On the other hand, it came from Google Research. 

So, combining these two factors I put my trust.

**Comment by sayakpaul** [comment]
@fchollet I have made the requested changes and also provided my explanations to your questions. PTAL when you have a moment.

File: examples/vision/learnable_resizer.py [hunk 23]
```diff
@@ -202,7 +201,7 @@ def get_model():
     backbone = tf.keras.applications.DenseNet(weights=None, include_top=True, classes=2)
```
Comment (s-mrb) [inline examples/vision/learnable_resizer.py:201]: I think it won't work, `tf.keras.application` has no function `DenseNet`

[densenet](https://www.tensorflow.org/api_docs/python/tf/keras/applications/densenet),

also when you use the given densenet variants you would have to use `input_shape` as

 ( `target_size[0], target_size[1],3)`

**Review by s-mrb — CHANGES_REQUESTED** [review]
I think `backbone` needs change!

File: examples/vision/learnable_resizer.py (hunk 23 above)
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py:201]: Yes. Thanks for pointing these out. Corrected.

File: examples/vision/learnable_resizer.py [hunk 24]
```diff
@@ -198,7 +198,12 @@ def learnable_resizer(
 
 
 def get_model():
-    backbone = tf.keras.applications.DenseNet(weights=None, include_top=True, classes=2)
+    backbone = tf.keras.applications.DenseNet121(
+        weights=None,
+        include_top=True,
+        classes=2,
+        input_shape=((INP_SIZE[0], INP_SIZE[1], 3)),
```
Comment (s-mrb) [inline examples/vision/learnable_resizer.py:205]: `input` of `backbone` is `output` of `learnable_resizer`, considering this the `input` of `backbone` must not be <br>
`((INP_SIZE[0], INP_SIZE[1], 3))`  instead it should be <br> `((TARGET_SIZE[0], TARGET_SIZE[1], 3))`

**Review by s-mrb — CHANGES_REQUESTED** [review]
Please check `input_shape`!
Correct me if i am wrong.

File: examples/vision/learnable_resizer.py (hunk 24 above)
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py]: Right. Corrected it.

File: examples/vision/learnable_resizer.py (hunk 18 above)
Comment (fchollet) [reply examples/vision/learnable_resizer.py]: That makes sense! Is this explained in the example sufficiently clearly?

File: examples/vision/learnable_resizer.py (hunk 19 above)
Comment (fchollet) [reply examples/vision/learnable_resizer.py]: Ok, sounds good.

File: examples/vision/learnable_resizer.py [hunk 25]
```diff
@@ -0,0 +1,285 @@
...
+
+"""
+The plot shows that the visuals of the images have improved with training. The following
+table shows the benefits of using the resizing module in comparison to using the bilinear 
+interpolation:
+
+|           Model           	| Number of  parameters (Million) 	| Top-1 accuracy 	|
+|:-------------------------:	|:-------------------------------:	|:--------------:	|
+|   With the learnable resizer  	|             7.051717            	|      67.67%     	|
+| Without the learnable resizer 	|             7.039554            	|      60.19%      	|
```
Comment (fchollet) [inline examples/vision/learnable_resizer.py:256]: These results are pretty convincing, the increase in parameter count is small and the increase in accuracy is solid!

File: examples/vision/learnable_resizer.py (hunk 18 above)
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py]: I can add these in the example briefly if you'd like.

File: examples/vision/learnable_resizer.py (hunk 25 above)
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py:256]: Yes, exactly. I did run the experiments for like three times and each time I was able to get a better performance with the resizer. To be absolutely sure and reproducible, I trained the models with the same initial random weights (I have mentioned it in the example).

File: examples/vision/learnable_resizer.py (hunk 18 above)
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py]: I added the following to clarify:

```
Now, a question worth asking here is -  _isn't the improved accuracy simply a consequence
of adding more layers (the resizer is a mini network after all) to the model, compared to
the baseline?_

To show that it is not the case, the authors conduct the following experiment:

* Take a pre-trained model trained some size, say (224 x 224).

* Now, first, use it to infer predictions on images resized to a lower resolution. Record
the performance.

* For the second experiment, plug in the resizer module at the top of the pre-trained
model and warm-start the training. Record the performance.

Now, the authors argue that using the second option is better because it helps the model
learn how to adjust the representations better with respect to the given resolution.
Since the results purely empirical, a few more experiments such as analyzing the
cross-channel interaction would have been even better. It is worth noting that elements
like [Squeeze and Excitation (SE) blocks](https://arxiv.org/abs/1709.01507), [Global Context (GC) blocks](https://arxiv.org/pdf/1904.11492) also add a few
parameters to an existing network but they are known to help a network process
information in systematic ways to improve the overall performance. 
```

**Comment by sayakpaul** [comment]
@fchollet I have added an extended explanation with regard to the results. PTAL and let me know if I am good to add the rest of the files.

File: examples/vision/learnable_resizer.py [hunk 26]
```diff
@@ -261,6 +261,28 @@ def get_model():
...
+
+* Now, first, use it to infer predictions on images resized to a lower resolution. Record
+the performance.
+
+* For the second experiment, plug in the resizer module at the top of the pre-trained
+model and warm-start the training. Record the performance.
+
+Now, the authors argue that using the second option is better because it helps the model
+learn how to adjust the representations better with respect to the given resolution.
+Since the results purely empirical, a few more experiments such as analyzing the
```
Comment (s-mrb) [inline examples/vision/learnable_resizer.py:281]: ..  `are` purely empirical ..

File: examples/vision/learnable_resizer.py [hunk 27]
```diff
@@ -261,6 +261,28 @@ def get_model():
 parameters due to the resizing module is very negligible. To ensure that the improvement
 in the performance is not due to stochasticity, the models were trained using the same
 initial random weights. 
+
+Now, a question worth asking here is -  _isn't the improved accuracy simply a consequence
+of adding more layers (the resizer is a mini network after all) to the model, compared to
+the baseline?_
+
+To show that it is not the case, the authors conduct the following experiment:
```
Comment (s-mrb) [inline examples/vision/learnable_resizer.py:269]: `conducted`

**Review by s-mrb — COMMENTED** [review]
Check these,

File: examples/vision/learnable_resizer.py (hunk 27 above)
Comment (sayakpaul) [reply examples/vision/learnable_resizer.py:269]: It's okay to write that part in present tense much like the publication.

**Comment by fchollet** [comment]
Looks great! Please add the generated files.

**Comment by sayakpaul** [comment]
Done @fchollet. Over to you.

**Review by fchollet — APPROVED** [review]
Thank you again for the great example! 👍
